### PR TITLE
Homework1

### PR DESCRIPTION
**Pull request by Magauiya**
## Checklist
- [x] Overall description of tech. decisions
- [x] Make checklist for PR
- [x] EDA and model prototype
- [x] Implement training script
- [x] Implement test script (inputs: model ckpt, testset, save path)
- [x] Module-based structure
- [x] Add logging
- [x] Test coverage: train, predict, etc.
- [x] Generate synthetic tests
- [x] Use config files (at least 2), additional points for using hydra
- [x] Use dataclasses, not dict
- [x] Freeze all dependencies
- [ ] Setup CI tests
- [ ] MlFlow: setup
- [ ] MlFlow: logging metrics     
- [ ] MlFlow: use Model Registry and add screenshot of run DVC usage
- [ ] MlFlow: choose entrypoints
- [ ] MlFlow: dataset versioning
- [ ] MlFlow: dvc pipeline  

As a solution for the heart disease classification, I chose LightGBM model. LightGBM is one of the SOTA in
terms of table data fitting. However, boosting models suffer from overfitting, therefore, I decreased depth, number of leafs,
and estimators. Also, training data is processed such that highly relied features are dropped. By this, we reach 80% accuracy 
compared to 71% accuracy of a default LightGBM.

### CODE REVIEW

File: ml_project/main.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    def train(self):
+        """
+        Train script
+        """
+        self.logger.debug("Loading dataset ...")
+        data = self._load_data(path=self.cfg.path.data)
+        x_train, x_test, y_train, y_test = train_test_split(
+            data.drop(["condition"], axis=1),
+            data["condition"],
```
Comment (SVizor42) [inline ml_project/main.py:55]: Don't hardcode target or feature names in the code. Please, use config files or constants for such purposes.

File: ml_project/main.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    def _save_ckpt(self):
+        save_path = os.path.join(
+            self.cfg.path.models, self.cfg.params.exp_name + ".pkl"
+        )
+        with open(save_path, "wb") as file:
+            pickle.dump(self.model, file)
+
+    def _preprocess(self, data):
+        drops = ["thalach", "chol", "age", "oldpeak", "trestbps", "thal"]
```
Comment (SVizor42) [inline ml_project/main.py:83]: Don't hardcode target or feature names in the code. Please, use config files or constants for such purposes.

File: ml_project/main.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        logFileFormatter = logging.Formatter(
+            fmt="%(levelname)s %(asctime)s Func: %(funcName)s - %(message)s",
+            datefmt="%Y-%m-%d %H:%M:%S",
+        )
+        filename = os.path.join(self.cfg.path.logs, f"train_{self.cfg.params.exp_name}.log")
+        fileHandler = logging.FileHandler(filename=filename)
+        fileHandler.setFormatter(logFileFormatter)
+        self.logger.addHandler(fileHandler)
+
+    def train(self):
```
Comment (SVizor42) [inline ml_project/main.py:47]: Return value type?

File: ml_project/main.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    def _preprocess(self, data):
+        drops = ["thalach", "chol", "age", "oldpeak", "trestbps", "thal"]
+        columns = data.columns
+        for col in drops:
+            if col in columns:
+                data = data.drop(columns=[col])
+        return data
+
+    def _load_data(self, path):
```
Comment (SVizor42) [inline ml_project/main.py:90]: Return value type?

File: ml_project/tests/test_fitpred.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import os
+import sys
+import unittest
+import omegaconf
+import pandas as pd
+from sklearn.metrics import accuracy_score
```
Comment (SVizor42) [inline ml_project/tests/test_fitpred.py:6]: The same

File: ml_project/tests/test_fitpred.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import os
+import sys
+import unittest
+import omegaconf
+import pandas as pd
```
Comment (SVizor42) [inline ml_project/tests/test_fitpred.py:5]: imported but unused

File: ml_project/main.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    def test(self):
+        """
+        Inference script
+        """
+        with open(self.cfg.test.ckpt, "rb") as file:
+            model = pickle.load(file)
+        data = self._load_data(path=self.cfg.test.data)
+        data = self._preprocess(data)
+        preds = model.predict(data)
+        pd.DataFrame(preds).to_csv(self.cfg.path.savefile)
```
Comment (SVizor42) [inline ml_project/main.py:109]: You need to check if this path exists or not.

File: ml_project/main.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        data = pd.read_csv(path)
+        if self.cfg.data.preprocess:
+            data = self._preprocess(data)
+        return data
+
+    def _make_dir(self):
+        os.makedirs(self.cfg.path.logs, exist_ok=True)
+        os.makedirs(self.cfg.path.models, exist_ok=True)
+
+    def test(self):
```
Comment (SVizor42) [inline ml_project/main.py:100]: Please, add some logging stuff for the inference part

**Review by SVizor42 — APPROVED** [review]
Well done.

Some things I want to emphasize:
1. Module based structure means that you split your code into separate modules (dataset building, feature preprocessing, training etc.). Ok, here you decided to define the single model class with appropriate methods. Looks legit. So, in this case you need to describe such technical solutions in the PR description.
2. I didn't find the dataclasses in the solution. Sorry, can't take this point into account.
3. Synthetic data - here you need to provide not only text files, but also the code\scripts that produced it. Just 1 point for this item.

**Sum: 24 points.**

You can merge, if you don't have any objections.

P.S. Please, add also `hw1` label to this PR.
